### ISSUE TITLE
Issues when closing / opening producers in Firefox

### ISSUE THREAD

**Issue by vpavlov-prime** [versatica/mediasoup-client#104]
With the update of mediasoup to version 3.3+ we started to experience issues with the FF browser (tested with version 71 on Windows). It turns out that when mediasoup-client is doing the webrtc handshaking it sets the SDP to use only the first sent track (mic or cam). I suppose that there is a bug in the mediasoup-client (we got the same results with 3.3.3 and 3.3.8) code. We have two different cases:
- When the user first start the mic, then stop it and after that try to start the cam they get:
`Error: Offer and answer have different number of m-lines (3 vs 2).`
- When the user first start the cam, then stop it and after that try to start the mic they get:
`Error: no a=ssrc lines found.`
</code></pre>
- The first created producer (no matter mic or cam) always succeeds, also if we keep the first producer started and then start the other one (no matter which is first), everything is working.
- These errors are not reproducible on Chrome.

These issues are not observed in the mediasoup-demo application, because the mic producer is started on join and then mute is used. In order to reproduce this in mediasoup-demo app a small change is needed - to use `disableMic()` instead of `muteMic()`. You can check my fork here:
https://github.com/vpavlov-prime/mediasoup-demo/commit/ebdea844a32400f7096a01a4f4377f44ba878562

**Note:** When using mediasoup 3.2.5 and mediasoup-client 3.2.5, the issue is not observed.

**Comment by ibc** [comment]
Cool. In bullet second in your issue description, Firefox is producing the following local SDP offer when calling to `pc.createOffer()`:

```
v=0
o=mozilla...THIS_IS_SDPARTA-71.0 6409989875250438094 6 IN IP4 0.0.0.0
s=-
t=0 0
a=fingerprint:sha-256 BA:A2:F6:F7:CF:D4:89:3C:84:B5:A4:23:F1:E0:D3:34:88:E1:7E:92:B9:98:75:60:10:F5:86:5C:70:A6:0E:A7
a=group:BUNDLE 0 3
a=ice-options:trickle
a=msid-semantic:WMS *
m=audio 52379 UDP/TLS/RTP/SAVPF 109 9 0 8 101
c=IN IP4 192.168.1.34
a=candidate:0 1 UDP 2122252543 192.168.1.34 52379 typ host
a=candidate:1 1 UDP 2122187007 5.5.5.5 58531 typ host
a=candidate:2 1 TCP 2105524479 192.168.1.34 9 typ host tcptype active
a=candidate:3 1 TCP 2105458943 5.5.5.5 9 typ host tcptype active
a=inactive
a=end-of-candidates
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2/recvonly urn:ietf:params:rtp-hdrext:csrc-audio-level
a=extmap:3 urn:ietf:params:rtp-hdrext:sdes:mid
a=fmtp:109 maxplaybackrate=48000;stereo=1;useinbandfec=1
a=fmtp:101 0-15
a=ice-pwd:6245d2c1b492bb5187eaabc41ba4ea05
a=ice-ufrag:ab3213b7
a=mid:0
a=rtcp-mux
a=rtpmap:109 opus/48000/2
a=rtpmap:9 G722/8000/1
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:101 telephone-event/8000/1
a=setup:actpass
a=ssrc:1445724942 cname:{31a21a1d-bb6f-9042-ab44-be863e7d8974}
m=video 0 UDP/TLS/RTP/SAVPF 120
c=IN IP4 0.0.0.0
a=inactive
a=mid:1
a=rtpmap:120 VP8/90000
m=audio 0 UDP/TLS/RTP/SAVPF 109 9 0 8 101
c=IN IP4 0.0.0.0
a=bundle-only
a=sendonly
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2/recvonly urn:ietf:params:rtp-hdrext:csrc-audio-level
a=extmap:3 urn:ietf:params:rtp-hdrext:sdes:mid
a=fmtp:109 maxplaybackrate=48000;stereo=1;useinbandfec=1
a=fmtp:101 0-15
a=ice-pwd:6245d2c1b492bb5187eaabc41ba4ea05
a=ice-ufrag:ab3213b7
a=mid:3
a=msid:{70fe393a-c2ba-0b42-982e-4a842870a413} {9bf81def-80df-e441-b76b-374e1b917af0}
a=rtcp-mux
a=rtpmap:109 opus/48000/2
a=rtpmap:9 G722/8000/1
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:101 telephone-event/8000/1
a=setup:actpass
a=ssrc:1355784695 cname:{31a21a1d-bb6f-9042-ab44-be863e7d8974}
```

which contains a `m=video` section with audio codecs. This is broken in Firefox. Will try to figure out how to avoid it in mediasoup-client but this is a bug that should be reported also in Firefox.

**Comment by ibc** [comment]
Mmm, sorry, I was wrong. The `m=video` section is disabled (media port 0) and there is a proper `m=audio` section after it.

**Comment by vpavlov-prime** [comment]
Please note, that there are two different errors, depending on which producer was started first.

**Comment by ibc** [comment]
Please, try the branch "issue-104" and let me know if this fixes both issues above (it does for me).

BTW it can be directly tested in the mediasoup-demo (no need for change anything in the code) via console:

```js
async function run1() {
	await CC.disableMic();
	await CC.enableWebcam();
}

run1();


async function run2() {
	await CC.disableMic();
	await CC.enableWebcam();
	await CC.disableWebcam();
	await CC.enableMic();
}

run2();
```

**Comment by vpavlov-prime** [comment]
Great work! It seams `issue-104` branch fixes the FF issues. I have tested with mediasoup-demo and in our project.

Thanks for the quick response!

**Comment by ibc** [comment]
Thanks. Released in mediasoup-client 3.3.12
